core/trino-main/src/main/java/io/trino/operator/window/matcher/Matcher.java

```diff
@@ -13,89 +13,100 @@
         private final Captures captures;
 
-        // For each instruction of the program, track if it was processed, and in which step.
-        private final int[] steps;
-
         public Runtime(Program program, int inputLength, boolean matchingAtPartitionStart)
         {
-            int size = program.size();
-            threads = new int[2 * size];
-            freeThreadIds = new IntStack(threads.length);
-            for (int i = threads.length - 1; i >= 0; i--) {
-                freeThreadIds.push(i);
-            }
-            this.captures = new Captures(threads.length, program.getMinSlotCount(), program.getMinLabelCount());
+            int initialCapacity = 2 * program.size();
+            threads = new IntMap(initialCapacity);
+            freeThreadIds = new IntStack(initialCapacity);
+            this.captures = new Captures(initialCapacity, program.getMinSlotCount(), program.getMinLabelCount());
             this.inputLength = inputLength;
-            this.steps = new int[size];
             this.matchingAtPartitionStart = matchingAtPartitionStart;
         }
 
-        public void nextStep()
-        {
-            step++;
-        }
-
-        public boolean hasCapacity(int size)
-        {
-            return freeThreadIds.size() >= size;
-        }
-
         private int forkThread(int parent)
         {
-            int threadId = freeThreadIds.pop();
-            captures.copy(parent, threadId);
-            return threadId;
+            int child = newThread();
+            captures.copy(parent, child);
+            return child;
         }
 
         private int newThread()
         {
-            int threadId = freeThreadIds.pop();
-            captures.allocate(threadId);
-            return threadId;
+            if (freeThreadIds.size() > 0) {
+                return freeThreadIds.pop();
+            }
+            return newThreadId++;
         }
 
         private void killThread(int threadId)
         {
             freeThreadIds.push(threadId);
             captures.release(threadId);
         }
+
+        private long getEstimatedSizeInBytes()
+        {
+            // estimation
+            long threadsAtInstructionsSize = SizeOf.estimatedSizeOf(
+                    threadsAtInstructions.asMap(),
+                    key -> INTEGER_SIZE,
+                    value -> SizeOf.sizeOfObjectArray(value.size()) + value.size() * (long) INTEGER_SIZE);
+
+            return INSTANCE_SIZE + threadsAtInstructionsSize + threads.getSizeInBytes() + freeThreadIds.getSizeInBytes() + captures.getSizeInBytes();
+        }
     }
 
     public Matcher(Program program)
     {
         this.program = program;
+        this.threadEquivalence = new ThreadEquivalence(program);
     }
 
-    public MatchResult run(LabelEvaluator labelEvaluator)
+    public MatchResult run(LabelEvaluator labelEvaluator, LocalMemoryContext memoryContext)
     {
+        threadEquivalence.setLabelEvaluator(labelEvaluator);
```

Comment:
We should figure out how to make this part of ThreadEquivalence at construction time. The information needed is static for the given program.